### ISSUE TITLE
Design Review - Abstract out Cryptographic Primatives

### ISSUE THREAD

**Issue by a-bash** [grempe/secretsharing#2]
This is refactoring idea I've been kicking around. I would be happy to
implement this if I could get your thoughts before I take the plunge.

Right now the implementation is coupled tightly with OpenSSL. This this makes
sense for most users but it does not seem to be a core concern of the
SSS Algorithm.

Specifically, I do not us OpenSSL in my project, I use DJB's NaCl. Also, maintaining
tight control over my Random Number generator, is very important to me.

I'm wondering if we can abstract away some of these more primative functions.
Something like:

```
module SecretSharing
  class BigNumber
    # call OpenSSL::BN from One Place
  end
end
```

```
module SecretSharing
  class RandomNumber
    # call OpenSSL Random Number Generations from One Place
  end
end
```

This would allow us to [depend on abstraction](http://en.wikipedia.org/wiki/Dependency_inversion_principle) and provide the opportunity to swap out crypto libraries, leaving OpenSSL as the smart default.

**Comment by grempe** [comment]
Hi Andrew, yes I think that makes sense. If you want to take a crack at it
that would be helpful.

I'm not yet sure if the math parts of the library can be done effectively
without openssl and just using stdlib. That would be good if it could
though.

Glenn

## 

Glenn Rempe

email                 : glenn@rempe.us
voice                 : (415) 894-5366
twitter                : @grempe
contact info        : http://www.rempe.us/contact
pgp                    : http://www.rempe.us/gnupg

**Comment by grempe** [comment]
OpenSSL has now been completely re-factored out of the codebase for 'secretsharing'. Sorry it took so long.  I just pushed a series of commit (which are likely backwards incompatible) which will become a new release in the near future.  I implemented the math functions that OpenSSL::BN was providing in pure Ruby. Consider this experimental for now.  I will likely also be changing the serialization format.

https://github.com/grempe/secretsharing/commit/0f739246be17f8fa6ed88e8c31166423e661a786

I think this finally closes this issue. Let me know if you have any feedback.

Cheers.
